### ISSUE TITLE
Parser incorrectly swaps BigQuery struct grammars when `parse_grammar` is not supplied.

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#3277]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When a `STRUCT` is used in conjunction with an `AS` alias, it's columns are recognised as datatypes, which can trigger L063.

### Expected Behaviour

`STRUCT` column names should not trigger L063.

### Observed Behaviour

```
L:   3 | P:   9 | L063 | Datatypes must be upper case.
L:   4 | P:   9 | L063 | Datatypes must be upper case.
```

### How to reproduce

Run `sqlfluff lint` on

```sql
SELECT
    STRUCT(
        some_field,
        some_other_field
    ) AS col,
    STRUCT(
        some_field,
        some_other_field
    )
FROM table
```

### Dialect

bigquery

### Version

0.13.1

### Configuration

```cfg
[sqlfluff]
templater = jinja
dialect = bigquery

[sqlfluff:rules:L063]
# Inconsistent capitalisation of datatypes
extended_capitalisation_policy = upper
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jmc-bbk** [comment]
Just doing a bit of investigation on this - although not really familiar with BigQuery or Ansi.

It's not to do with the `alias` segment. You will find the same result with:

```
SELECT
    STRUCT(
        some_field,
        some_other_field
    ),
    STRUCT(
        some_field,
        some_other_field
    )
FROM table
```

Going to try figure it out - but anyone else welcome to jump on the ticket!

**Comment by barrywhart** [comment]
Yes, the parser seems to be behaving inconsistently. For this query, the first 3 `STRUCT` parameters parse as `data_type_identifier`, but the 4th parses as `column_reference`:
```
SELECT
    STRUCT(field1, field2) AS col1,
    STRUCT(field1, field2),
    STRUCT(field1, field2) AS col2,
    STRUCT(field1, field2)
FROM table
```

Here, all 6 parse as `data_type_identifier`:
```
SELECT
    STRUCT(field1, field2),
    STRUCT(field1, field2),
    STRUCT(field1, field2),
    STRUCT(field1, field2),
    STRUCT(field1, field2),
    STRUCT(field1, field2),
FROM table
```

@tunetheweb, @WittierDinosaur: Any ideas about this? This is definitely a parser or dialect issue, not a rule issue. Perhaps the dialect is ambiguous? Unsure.

**Comment by barrywhart** [comment]
Trailing commas make a difference. Same query as above, but added a trailing comma to the column list. Now all are `data_type_identifier`:
```
SELECT
    STRUCT(field1, field2),
    STRUCT(field1, field2),
FROM table
```

But same query, if I remove the `FROM table`, now they are all `column_reference`.
```
SELECT
    STRUCT(field1, field2),
    STRUCT(field1, field2),
```

**Comment by tunetheweb** [comment]
Had a quick look at this last night and couldn’t figure it out. It’s definitely ambiguous but not sure why it’s picking on for one, and another for another.

Will have another look later in the weeekend if no one else figures it out first.

**Comment by jmc-bbk** [comment]
I think the issue ultimately lies with:

```
class DatatypeSegment(ansi.DatatypeSegment):
    """A data type segment.

    In particular here, this enabled the support for
    the STRUCT datatypes.
    """

    match_grammar = OneOf(  # Parameter type
        Ref("DatatypeIdentifierSegment"),  # Simple type
        Sequence("ANY", "TYPE"),  # SQL UDFs can specify this "type"
        Ref("SimpleArrayTypeGrammar"),
        Ref("StructTypeGrammar"),
    )
```

if I comment this piece of code out and run the parse, I get expected results.

**Comment by WittierDinosaur** [comment]
The number of fields inside the bracketed struct also make no difference
```
SELECT
    STRUCT(field1),
    STRUCT(field1)
FROM table
```

**Comment by WittierDinosaur** [comment]
Also, if you change the definition of typeless struct to:
```
class TypelessStructSegment(ansi.TypelessStructSegment):
    """Expression to construct a STRUCT with implicit types.

    https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#typeless_struct_syntax
    """

    match_grammar = Sequence(
        "STRUCT",
        Bracketed(
            Ref("ColumnReferenceSegment"),
            ),
    )
```
It still somehow manages to conjure a data_type. Confirmed @jmc-bbk is right on commenting out that code working - but I have no idea how it's even getting to data_type

**Comment by tunetheweb** [comment]
Change BaseExpressionType to replace ExpressionSegment with TypelessStructSegment also fixes it. Not not really what we want to be honest.

ColumnReferenceSegment is also overridden in BigQuery but still couldn’t figure out how it got there either and gave up for the day yesterday to look with fresh eyes at some point.

**Comment by WittierDinosaur** [comment]
I wonder if this is to do with it behaving differently in match and parse? I don't have a whole lot of time to look at this today, but may have a crack tomorrow if nobody solves. I'm more concerned about the behaviour problem than the use-case here. It feels like something is leaking

**Comment by tunetheweb** [comment]
I think you're right. Adding an explicit `parse_grammar` avoids the issue:

```py
class TypelessStructSegment(ansi.TypelessStructSegment):
    """Expression to construct a STRUCT with implicit types.

    https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#typeless_struct_syntax
    """

    match_grammar = Sequence(
        "STRUCT",
        Bracketed(
            Delimited(
                Sequence(
                    Ref("BaseExpressionElementGrammar"),
                    Ref("AliasExpressionSegment", optional=True),
                ),
            ),
        ),
    )

    parse_grammar = Sequence(
        "STRUCT",
        Bracketed(
            Delimited(
                Sequence(
                    Ref("BaseExpressionElementGrammar"),
                    Ref("AliasExpressionSegment", optional=True),
                ),
            ),
        ),
    )
```

Question is - why?

**Comment by barrywhart** [comment]
So there's currently a segment definition with a `match_grammar` but not a `parse_grammar`? That seems wrong -- if you only have one, shouldn't it be `parse_grammar`? Is inheritance from the base dialect part of the issue here?

**Comment by tunetheweb** [comment]
No you’ve got it the wrong way around @barrywhart - it’s allowed not to specify a `parse_grammar`, but in that case the `match_grammar` is used as the `parse_grammar`.

So there should be no need to specify it separately and it should parse in the same way even without it.

it’s odd…

**Comment by tunetheweb** [comment]
https://docs.sqlfluff.com/en/stable/architecture.html#stage-3-the-parser

> Each segment can specify a parse_grammar, and a match_grammar. If both a match_grammar and parse_grammar are defined in a segment, match_grammar is used to quickly prune the tree for branches which do not match segments being parsed, and the parse_grammar is then used to refine the branch identified as correct. If only a match_grammar is defined, then it serves the purpose of both pruning and refining.

**Comment by tunetheweb** [comment]
It's somehow confusing `StructTypeGrammar` and `TypelessStructSegment` - both of which start with `STRUCT` but then have different brackets.

**Comment by dmohns** [comment]
This issue is extremely weird. But I can confirm that setting a `parse_grammar` on `TypelessStructSegment` seems to resolve this issue.

What do you think about setting the `parse_grammar` for now to get rid of the false positives. We can continuing to investigate if there is some issue with the parser based on commit prior to the fix?

**Comment by tunetheweb** [comment]
And let’s add a comment explaining why both are needed for now so someone like me doesn’t clean it up later 😁

**Comment by tunetheweb** [comment]
I suspect this is related to: https://github.com/sqlfluff/sqlfluff/pull/997 (was reminded of this recently).

And I wonder if, when we figure out the actual fix for this, that workaround can be removed too?

**Comment by barrywhart** [comment]
Doing some legwork before taking a close look...

With coverage tracking enabled, I ran the test added in PR #3306 both with and without the workaround. Below is a partial list of the modules whose coverage differed from one run to the other. (I omitted some that seemed unlikely to be the location of the bug.)

Next step: Look at line-level coverage info for these modules, probably focusing on `sqlfluff.core.parser` first.

```
< src/sqlfluff/core/parser/markers.py                                        70      7    90%
> src/sqlfluff/core/parser/markers.py                                        70      6    91%

< src/sqlfluff/core/parser/segments/base.py                                 547    211    61%
> src/sqlfluff/core/parser/segments/base.py                                 547    212    61%

< src/sqlfluff/core/parser/segments/raw.py                                  116     18    84%
> src/sqlfluff/core/parser/segments/raw.py                                  116     17    85%

< src/sqlfluff/core/errors.py                                                68     30    56%
> src/sqlfluff/core/errors.py                                                68     18    74%

< src/sqlfluff/core/linter/linter.py                                        403    205    49%
> src/sqlfluff/core/linter/linter.py                                        403    201    50%

< src/sqlfluff/core/rules/base.py                                           426    183    57%
> src/sqlfluff/core/rules/base.py                                           426    127    70%

< src/sqlfluff/core/templaters/base.py                                      166     41    75%
> src/sqlfluff/core/templaters/base.py                                      166     28    83%

```

**Comment by barrywhart** [comment]
Line-level, fails on left, works on right (commit `cb6357c540d2d968f766f3a7a4fa16f231cb80e4`):
```
< src/sqlfluff/core/parser/markers.py                  70      6    91%   66, 70, 124, 129, 158, 162
> src/sqlfluff/core/parser/markers.py                  70      7    90%   66, 70, 124, 129, 158, 162, 207

< src/sqlfluff/core/parser/segments/base.py           547    212    61%   68-71, 91, 95, 110-111, 116, 126-135, 249, 265, 301, 305, 352, 444, 493, 546, 557, 559, 567-584, 698, 702, 710-713, 717, 721, 725, 729, 777-805, 813, 828-834, 843-846, 850, 854-857, 861-865, 879-887, 891-897, 917-918, 940-956, 989, 1035-1053, 1095, 1111-1256, 1261-1267, 1271-1284, 1311-1405, 1416, 1465-1467, 1479-1480, 1487, 1494, 1524, 1528-1531, 1546-1547, 1550, 1556, 1560
> src/sqlfluff/core/parser/segments/base.py           547    211    61%   68-71, 91, 95, 110-111, 116, 126-135, 232, 249, 265, 301, 352, 444, 493, 546, 557, 559, 567-584, 698, 702, 710-713, 717, 721, 725, 729, 777-805, 813, 828-834, 843-846, 850, 854-857, 861-865, 879-887, 891-897, 917-918, 940-956, 1035-1053, 1095, 1111-1256, 1261-1267, 1271-1284, 1311-1405, 1416, 1465-1467, 1479-1480, 1487, 1494, 1524, 1528-1531, 1546-1547, 1550, 1556, 1560
```
< src/sqlfluff/core/parser/segments/raw.py            116     17    85%   120, 134, 138-154, 279
> src/sqlfluff/core/parser/segments/raw.py            116     18    84%   120, 134, 138-154, 181, 279

**Comment by barrywhart** [comment]
I'm pretty sure the bug manifests in `src/sqlfluff/core/parser/segments/base.py`. There are 3 differences in lines covered between the failing vs working cases:
* 232
* 305
* 989

The second and third lines above relate to expanding of segments:
```
304        elif self.segments and any(s.is_expandable for s in self.segments):
305            return True 
```

```
987        if parse_grammar is None:
988            # No parse grammar, go straight to expansion
989            parse_context.logger.debug(
990                "{}.parse: no grammar. Going straight to expansion".format(
991                    self.__class__.__name__
992                )
993            )
```

@tunetheweb: This is exactly the area we were discussing a couple of weeks ago in my PR #3407, "Fix bug in apply_fixes() with leading/trailing whitespace".

@alanmcruickshank: Any ideas on how we might try to fix this?

**Comment by barrywhart** [comment]
I created a PR #3427 which is a naive attempt to fix it (falls back to `match_grammar` if there's no `parse_grammar`).

**Comment by alanmcruickshank** [comment]
Thanks for all the digging everyone (and thanks for the tag @barrywhart). I definitely agree that something is up here in the parser.

@barrywhart - I'm going to do some proper debugging on this tomorrow. I don't think we should need to do what you've suggested in #3427, because the `match_grammar` should _already have been executed_ on the segment to have matched in the first place. In theory it shouldn't need to be run so something else is going on here.

**Comment by alanmcruickshank** [comment]
I'm pretty sure I know what's happening here. The section is being correctly parsed by `TypelessStruct`, but somehow is _afterwards_ being reparsed by `StructTypeGrammar` which will _also_ match. The latter is a grammar and not a segment, so doesn't mutate the segment name - but it _is_ mutating the segment types.

Working on a solution now.

**Comment by tunetheweb** [comment]
Note it’s not an exact match. The bracket types are different in both.

**Comment by alanmcruickshank** [comment]
I'm still working on this - I've narrowed it down to a section of `Bracketed` where we're assuming something is treated as immutable - but it's very much not. I expect that I'll have some more precise tests which come out of the fix when I find it.

**Comment by alanmcruickshank** [comment]
Ok - draft PR is up #3428. There was a mutation happening during `.match()` functions which meant it effectively had side effects when it shouldn't.

It makes me feel like we might need to do something more drastic in enforcing immutability in some of these places. The PR also doesn't currently include any tests - I'm open to suggestions on how we best want to test this.

**Comment by alanmcruickshank** [comment]
I'm going to mark this issue as resolved now that #3590 and #3428 are resolved.
